### PR TITLE
KAFKA-10790 Detect/Prevent Deadlock on Producer Network Thread

### PR DESCRIPTION
**Pull request by g1geordie**
KafkaProducer.flush method in callback will cause deadlock .

because flush method  wait the future complete
``` java
//  KafkaProducer
public void flush() {
   .....
        try {
            this.accumulator.awaitFlushCompletion();
        } catch (InterruptedException e) {
            throw new InterruptException("Flush interrupted.", e);
        }
}

//RecordAccumulator
    public void awaitFlushCompletion() throws InterruptedException {
        try {
            for (ProducerBatch batch : this.incomplete.copyAll())
                batch.produceFuture.await();
        } finally {
            this.flushesInProgress.decrementAndGet();
        }
    }
```

but  future complete after the callback .
```java
//ProducerBatch
 private void completeFutureAndFireCallbacks(long baseOffset, long logAppendTime, RuntimeException exception) {
     ...
        // execute callbacks
        for (Thunk thunk : thunks) {
            try {
                if (exception == null) {
                    RecordMetadata metadata = thunk.future.value();
                    if (thunk.callback != null)
                        thunk.callback.onCompletion(metadata, null);
                } else {
                    if (thunk.callback != null)
                        thunk.callback.onCompletion(null, exception);
                }
            } catch (Exception e) {
                log.error("Error executing user-provided callback on message for topic-partition '{}'", topicPartition, e);
            }
        }

        produceFuture.done();
 }
```

### CODE REVIEW

**Comment by g1geordie** [comment]
@chia7712 Can you help me review?

File: clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java [hunk 1]
```diff
@@ -1115,6 +1115,10 @@ private void ensureValidRecordSize(int size) {
      */
     @Override
     public void flush() {
+        boolean invokedFromCallback = Thread.currentThread() == this.ioThread;
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java]: As this check is used by ```close``` also, could you add a helper method to return boolean?

File: clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java [hunk 2]
```diff
@@ -391,6 +391,23 @@ Sender newSender(LogContext logContext, KafkaClient kafkaClient, ProducerMetadat
...
+        Map<String, Object> configs = new HashMap<>();
+        configs.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:9999");
+        ProducerMetadata metadata = mock(ProducerMetadata.class);
+
+        when(metadata.fetch()).thenReturn(emptyCluster, emptyCluster, emptyCluster, onePartitionCluster);
+
+        KafkaProducer<String, String> producer = producerWithOverrideNewSender(configs, metadata);
+        ProducerRecord<String, String> record = new ProducerRecord<>(topic, "value");
+
+        producer.send(record, (metadata1, exception) -> producer.flush());
```
Comment (chia7712) [inline clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java]: Could you add a flag after ```producer.flush()```? We should make sure ```producer.flush()``` fails.

Comment (g1geordie) [reply clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java]: Thank you  . I add a Assert.fail after flush

File: clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java [hunk 3]
```diff
@@ -1079,6 +1079,13 @@ private void ensureValidRecordSize(int size) {
                     " configuration.");
     }
 
+    /**
+     * @return is called in callback.
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java]: I feel the method name is good to be document.

File: clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java [hunk 4]
```diff
@@ -1201,7 +1211,7 @@ private void close(Duration timeout, boolean swallowException) {
 
         // this will keep track of the first encountered exception
         AtomicReference<Throwable> firstException = new AtomicReference<>();
-        boolean invokedFromCallback = Thread.currentThread() == this.ioThread;
+        boolean invokedFromCallback = isInvokedFromCallback();
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java]: this local variable is redundant

File: clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java [hunk 5]
```diff
@@ -391,6 +391,27 @@ Sender newSender(LogContext logContext, KafkaClient kafkaClient, ProducerMetadat
...
+        ProducerMetadata metadata = mock(ProducerMetadata.class);
+
+        when(metadata.fetch()).thenReturn(emptyCluster, emptyCluster, emptyCluster, onePartitionCluster);
+
+        KafkaProducer<String, String> producer = producerWithOverrideNewSender(configs, metadata);
+        ProducerRecord<String, String> record = new ProducerRecord<>(topic, "value");
+
+        producer.send(record, (metadata1, exception) -> {
+            producer.flush();
+            Assert.fail("Flush method needs to throw exception.");
```
Comment (chia7712) [inline clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java]: I don't think this is valid since this callback is triggered by another thread. The assert failure can't be aware by test thread.

**Review by chia7712 — COMMENTED** [review]
@g1geordie Thanks for your patch. Some comments are left. Please take a look.

File: clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java (hunk 3 above)
Comment (g1geordie) [reply clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java]: remove

File: clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java (hunk 4 above)
Comment (g1geordie) [reply clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java]: direct to method

File: clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java (hunk 5 above)
Comment (g1geordie) [reply clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java]: move assert to main thread

**Review by g1geordie — COMMENTED** [review]
thanks for all suggestion .

File: clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java [hunk 6]
```diff
@@ -1115,6 +1119,9 @@ private void ensureValidRecordSize(int size) {
      */
     @Override
     public void flush() {
+        if (isInvokedFromCallback())
+            throw new UnsupportedOperationException("Can't specify flush method in callback.");
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java]: It seems to me ```IllegalStateException``` is more suitable for this case. Also, please update docs of ```flush()```

File: clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java [hunk 7]
```diff
@@ -391,6 +391,31 @@ Sender newSender(LogContext logContext, KafkaClient kafkaClient, ProducerMetadat
...
+        Map<String, Object> configs = new HashMap<>();
+        configs.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:9999");
+        ProducerMetadata metadata = mock(ProducerMetadata.class);
+
+        when(metadata.fetch()).thenReturn(emptyCluster, emptyCluster, emptyCluster, onePartitionCluster);
+
+        KafkaProducer<String, String> producer = producerWithOverrideNewSender(configs, metadata);
+        ProducerRecord<String, String> record = new ProducerRecord<>(topic, "value");
+
+        AtomicBoolean  exceptionThrow = new AtomicBoolean(false);
```
Comment (chia7712) [inline clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java]: Could you save the exception caused by ```producer.flush()``` and then check the error type in the test thread?

**Review by chia7712 — COMMENTED** [review]
@g1geordie Thanks for your updating. Two minor comments are left.

File: clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java [hunk 8]
```diff
@@ -391,6 +391,31 @@ Sender newSender(LogContext logContext, KafkaClient kafkaClient, ProducerMetadat
...
 
+    @Test(timeout = 10000)
+    public void testFlushInCallbackNotCauseDeadlock() throws InterruptedException {
+        Map<String, Object> configs = new HashMap<>();
+        configs.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:9999");
+        ProducerMetadata metadata = mock(ProducerMetadata.class);
+
+        when(metadata.fetch()).thenReturn(emptyCluster, emptyCluster, emptyCluster, onePartitionCluster);
+
+        KafkaProducer<String, String> producer = producerWithOverrideNewSender(configs, metadata);
```
Comment (chia7712) [inline clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java]: Could you use try block?

File: clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java [hunk 9]
```diff
@@ -1116,11 +1116,12 @@ private boolean isInvokedFromCallback() {
      * </p>
      *
      * @throws InterruptException If the thread is interrupted while blocked
+     * @throws IllegalStateException If flush method in callback
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java]: ```If the thread is callback thread. Calling flush() from callback is not allowed since it makes deadlock.```

File: clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java [hunk 10]
```diff
@@ -1116,11 +1116,12 @@ private boolean isInvokedFromCallback() {
...
      */
     @Override
     public void flush() {
         if (isInvokedFromCallback())
-            throw new UnsupportedOperationException("Can't specify flush method in callback.");
+            throw new IllegalStateException("Can't specify flush method in callback.");
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java]: ```Calling flush() from callback is not allowed since it makes deadlock.```

**Comment by g1geordie** [comment]
@chia7712  thanks again

File: clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java [hunk 11]
```diff
@@ -131,6 +131,14 @@
             Collections.singletonList(new PartitionInfo(topic, 0, null, null, null)),
             Collections.emptySet(),
             Collections.emptySet());
+
+    private final Cluster oneLeaderCluster = new Cluster(
```
Comment (chia7712) [inline clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java]: Pardon me, why this change is required.

**Review by chia7712 — APPROVED** [review]
@g1geordie Thanks for your patch. LGTM

File: clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java (hunk 11 above)
Comment (g1geordie) [reply clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java]: producer.close(Duration.of(0))  will force close without sending all message . 
producer.close()  will gracefully close after sending all message .

the callback method will be invoke after  one of above  does

when we want to  send message  , we need to have
-  A Cluster has a leader partition 
-  Mockclient  prepare a fake response

So I add a cluster with one leader

Comment (g1geordie) [reply clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java]: if the message is not sent  , the thread will have a infinite loop

```
#Sender 
while(  ... .  this.accumulator.hasUndrained() || this.client.inFlightRequestCount() > 0  ...)
```

Comment (chia7712) [reply clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java]: Thanks for your explanation. ```oneLeaderCluster``` is used by this test case only. Could you make it be a local variable?

Comment (g1geordie) [reply clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java]: you are so nice !

**Review by chia7712 — COMMENTED** [review]
Do other methods which need to wait for IO thread have similar issue? For example, calling ```sendOffsetsToTransaction``` in callback always causes the timeout as the IO thread is locked by the "waiting".

Is it worth adding similar check for them? A explicit exception about "this is disallowed in callback" should be more suitable than ```TimeoutException```.

**Comment by chia7712** [comment]
@hachikuji FYI

**Comment by ijuma** [comment]
Have we considered whether we could do something similar to `close` instead of throwing an exception? Not saying that's a good idea, just want to make sure we thought through the options.

**Comment by g1geordie** [comment]
@ijuma hello 

The close method want to prevent self-join but there are no deadlock  . 

The flush method  in callback has a deadlock.
The deadlock is because  `flush` wait the `sender` send all message .
but messages are done after callback complete .
**flush in callback  , callback will never complete .**

`flush`'s semanteme is send message . 
so wait all message send I think it's necessary .

Messages are done  after callback I think is option ,
Or we can make async to `callback`  ,then messages will done no matter `callback` is complete.

**Comment by chia7712** [comment]
Personally, the callback is a bit weird to me. The callback mechanism makes “internal” thread to handle user-defined code. The user code to handle response/error is vary different from serialization so it seems to me that callback causes some problems.

1. one thread’s callback can hurt other threads when we share a prouder to different threads.
2. bad performance due to “single” thread
3. can’t isolate the internal thread from user
4. callback has some cryptic limits (for instance, this issue and #9842)

I prefer the Future that user can handle it by user-defined thread. Also, it  does not have limits like callback and user code can be isolated from internal thread. For another, we should consider CompletabeFuture as we are already in JDK 8.

**Comment by chia7712** [comment]
close as @frankvicky is working on it
